### ISSUE TITLE
TypeError in sinon.js: 'toString' of undefined

### ISSUE THREAD

**Issue by kirrg001** [sinonjs/sinon#648]
Hey,

sinon version: 1.12.2
karma sinon: 1.0.4

Sinon works fine, but when i console log the window (console.log(window)) it throws the following error:

````
  Uncaught TypeError: Cannot read property 'toString' of undefined
  at /Users/kirrg/dev/avuba/avb-app-v2/node_modules/sinon/pkg/sinon.js:2138
````

`````
  toString: function () {
     var callStr = this.proxy.toString() + "(";
`````

this.proxy is undefined. how can i fix this?
i couldnt find a duplicate, if it is, just delete 'me'.

thank you!
kate

**Comment by mroderick** [comment]
@kirrg001 thank you for reporting this :+1: 

I haven't seen this error in a bug report before, so perhaps it is new :)

Would you mind providing a bit more detail on how you produced the error, perhaps a code snippet or a test file that can trigger it?

**Comment by kirrg001** [comment]
i could send you a zip file.
it contains the minimal structure to reproduce the error.

**Comment by mroderick** [comment]
> i could send you a zip file.
> it contains the minimal structure to reproduce the error.

If it's not sensitive, then [publishing it as a gist](https://gist.github.com) or similar would be preferable, that would allow people other than me to contribute to solving the issue

**Comment by kirrg001** [comment]
git@github.com:kirrg001/sinon-bug.git

**Comment by kirrg001** [comment]
just tell me if i can help you.
regards
kirrg001

**Comment by djforth** [comment]
+1 getting the same issue, have tracked the issue down to the call.js line154 it looks like this.proxy is returning undefined.  

``` javascript
var callStr = this.proxy.toString() + "(";

```

But not entirely sure why?  maybe an issue with using Mocha and Chai with sinon?

**Comment by djforth** [comment]
Ok I tried to figure out the issue, but haven't tracked it down any further. However it seems to be something with using Mocha with Sinon causes the issue, possibly with Karma (not sure about that - but that is the same stack that kirrg001 is using from what I can see). 

I've switch to using jasmine/sinon/karma and no issues, so must be something around that.  If get a chance to have another look I will. Hope that helps

**Comment by kirrg001** [comment]
any news?what can i do to prevent the problem?can i give you more information?

**Comment by mroderick** [comment]
There's RequireJS in there, have you tried using a pre-built Sinon?

https://github.com/cjohansen/Sinon.JS/#important-amd-needs-pre-built-version

**Comment by JohnBerlin** [comment]
I'm a

**Comment by JohnBerlin** [comment]
I'm also having this issue with 

"karma": "~0.12.0"
"karma-mocha": "~0.1"
"sinon": "^1.15.4"

I fixed it temporarly by modifying the built version of sinon 
```
var callStr = (this.proxy ? this.proxy.toString():"") + "(";
var args = [];

 if (!this.args){
   return ":(";
 }
```
but I have no idea what the impact could be (but it works fine)

**Comment by iainctduncan** [comment]
I'm having the same issue with the last two built releases. The hack above is fixing it for me too. Would be great to get a build release working for Require.js! thanks.

**Comment by aladine** [comment]
I have the same issue in sinon.js 1.14.1,  in IE11.

**Comment by fatso83** [comment]
I'll add some info I have found out. It seems, as others have pointed out, that this is somehow related to Karma (or Mocha?) interfering somehow with what SinonJS is doing, as adding a test with `console.log(window)` to a test suite run by Karma will blow up, while adding it to the test suite in Sinon that is run by BusterJS will run just fine. 

An interesting thing to see is that the output of console.log(window) shows that `window` is somehow wrapped by Sinon:
```
{
  match: function match() {},
  mock: function () {},
  sandbox: {  },
  spy: function () {},
  stub: function () {},
  useFakeServer: function () {},
  useFakeTimers: function () {}
}
```

That might at least answer why `console.log(sinon)` has exactly the same effect in a Karma test (where sinon has been loaded). Although I still have now idea what happens.

I made a test script that can show where this behaviour (error?) was introduced:
```
curl -o test-script.sh https://raw.githubusercontent.com/fatso83/Sinon.JS/console.log.bug/test-script.sh
git bisect start v1.7.0 v1.6.0   && git bisect run sh test-script.sh
```
The output shows that the change in behaviour was introduced in b978955. 
@mantoni: do you have an idea what is happening?

**Comment by joscha** [comment]
This error comes up for me as well independently of a `console.log(window)` on IE 11 only (all other browsers tested seem to be fine, incl. Edge):
```
IE 11.0.0 (Windows 7 0.0.0) aui/dialog2-integration focuses inside panel not on buttons FAILED
	TypeError: Unable to get property 'toString' of undefined or null reference
	   at callProto.toString (/Users/jfeth/dev/aui/node_modules/sinon/pkg/sinon.js:2334:17)
	   at module.exports (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:3946:3)
	   at formatValue (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4469:5)
	   at formatProperty (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4593:9)
	   at Anonymous function (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4508:7)
	   at formatValue (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4507:5)
	   at inspect (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4363:3)
	   at module.exports (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:4698:3)
	   at module.exports (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:3920:3)
	   at Assertion.prototype.assert (/Users/jfeth/dev/aui/node_modules/karma-sinon-chai/node_modules/chai/chai.js:204:7)
```
Any way we can have #833 merged to fix this?

**Comment by fatso83** [comment]
@joscha : This should be fixed in master now. Pull the latest and see if that works for you.

**Comment by joscha** [comment]
@fatso83 great, would be awesome if we could get a release of this!

**Comment by fatso83** [comment]
As 2.0 is not ready for release yet, maybe we will release a patch release of 1.17 first, but that requires a separate pr.

**Comment by joscha** [comment]
I did for testing, but for adding this, I need a release, because karma-sinon-chai uses the pkg'd version of sinon which is not part of a git install.

**Comment by fatso83** [comment]
Fixed in #970 and #833
